oak/server/oak_node.cc

```diff
@@ -125,8 +128,17 @@ bool OakNode::WaitOnChannels(std::vector<std::unique_ptr<ChannelStatus>>* status
       return false;
     }
 
-    // TODO: get rid of polling wait
-    absl::SleepFor(absl::Milliseconds(100));
+    auto notify = std::make_shared<Notification>();
+    for (MessageChannelReadHalf* channel : readable_channels) {
+      channel->AddNotify(std::weak_ptr<Notification>(notify));
+    }
```

Comment:
I think here we need to check for messages directly after adding ourselves as a reader. 

It can be the case that a message was added & notifications were triggered after we tried reading, but before we added ourselves to receive notifications. The timeout below would prevent it deadlocking forever, but it would be better to avoid.

Let me know what you think?

Reply:
Removed the synchronization-by-sleep, and shifted to a single Mutex protecting all per-channel state.